png-read-write.md

```diff
@@ -0,0 +1,105 @@
+	- reader.d
+	- writer.d
+
+##### Common.d
+
+It contains code common for reader and writer: structs that mirror the file records or are part of both APIs, definitions of chunk types and similar. The main API data structure is this:
+
+```d
+struct PngImage
+{
+    PngDataStatus status;
+    string error;
+    // Linked list of chunk data: char[4] type, ubyte[] data. Contains PLTE and ancillary chunks.
+    PngChunk* chunk;
+    ubyte[] pixels;
+    PngHeaderData info;
+    alias info this;
+}
+
+enum PngImageStatus
+{
+    empty,
+    infoLoaded, // Only the header data is loaded
+    dataLoaded, // All data is loaded
+    errorEncountered
+}
+
+// This structure mirrors the file record
+align(1)
+struct PngHeaderData
+{
+    uint width;
+    uint height;
+    PngBitDepth bitDepth;
+    PngColorType colorType;
+    PngCompressionMethod compressionMethod;
+    PngFilterMethod filterMethod;
+    PngInterlaceMethod interlaceMethod;
+}
+
+// This is used to specify the desired format of pixels after they are read from file and processed
+enum PngFormat {
+    raw = 0,
+    // The lower 4 bits tells how much bytes each format require per pixel
+    gray = 1,
+    grayAlpha = 2,
+    rgb = 3,
+    rgba = 4,
+    rgb16 = 6,
+    rgba16 = 8,
+    gray16 = 0x12,
+    grayAlpha16 = 0x14
+}
+```
+
+##### Reader.d
+
+API are the following functions:
+
+```d
+// Loads only the header info into the PngImage struct. Doesn't need allocator.
+PngImage loadPngHeader(ref ubyte[] file) nothrow @nogc;
+
+// Loads whole image into PngImage struct.
+PngImage loadPng(ref ubyte[] file, PngFormat desiredFormat = PngFormat.rgba);
+PngImage loadPng(ref ubyte[] file, Allocator* allocator, PngFormat desiredFormat = PngFormat.rgba) nothrow @nogc;
+```
+File array is given as ref because these functions will move it behind the read data which might be useful if loading from pak file or something like that. In case you called `loadPngHeader` and now have PngImage with only header loaded and you want to load its data you can use one of these:
+```d
+struct PngImage {
+    ...
+    // They return false on error but also update the status and error string in the structure
+    bool loadData(ref ubyte[] file, PngFormat desiredFormat = PngFormat.rgba);
+    bool loadData(ref ubyte[] file, Allocator* allocator, PngFormat desiredFormat = PngFormat.rgba) nothrow @nogc
+}
+```
+
+
+## Open Questions
+
+1. Supporting interlaced (progressive) format would bring in a lot of complexity, would probably require more API changes and it is very rare these days. Would we be ok with the decision not to support it until further notice? (Disclaimer: I am not implementing it either way :))
+2. Should we make a separate float gamma field in PngImage struct that gets populated from gAMA chunk automatically instead of having to look for it in chunk data linked list?
+3. Should we even use gamma from the file as given? This [article](https://hsivonen.fi/png-gamma/) speaks of problems with it.
```

Comment:
Lets not apply it. But we can still load it ;)

Reply:
We don't 'apply' it... it's just metadata.